### PR TITLE
Add LSP handler for copilot related documents 

### PR DESCRIPTION
**Pull request by CyrusNajmabadi**
Followup to https://github.com/dotnet/roslyn/pull/74906.

### CODE REVIEW

File: src/LanguageServer/Protocol/Microsoft.CodeAnalysis.LanguageServer.Protocol.csproj [hunk 1]
```diff
@@ -117,6 +117,10 @@
     </EmbeddedResource>
   </ItemGroup>
 
+  <ItemGroup>
+    <Folder Include="Handler\RelatedDocuments\" />
+  </ItemGroup>
```
Comment (genlu) [inline src/LanguageServer/Protocol/Microsoft.CodeAnalysis.LanguageServer.Protocol.csproj:122]: is this needed?

Comment (CyrusNajmabadi) [reply src/LanguageServer/Protocol/Microsoft.CodeAnalysis.LanguageServer.Protocol.csproj]: nope.  removed.

**Comment by CyrusNajmabadi** [comment]
@dibarbet @genlu this is ready for review.

File: src/LanguageServer/Protocol/Handler/RelatedDocuments/RelatedDocumentsHandler.cs [hunk 2]
```diff
@@ -0,0 +1,147 @@
...
+        var document = context.Document;
+        Contract.ThrowIfNull(solution);
+        Contract.ThrowIfNull(document);
+
+        context.TraceInformation($"Processing: {document.FilePath}");
+
+        var relatedDocumentsService = document.GetLanguageService<IRelatedDocumentsService>();
+        if (relatedDocumentsService == null)
+        {
+            context.TraceInformation($"Ignoring document '{document.FilePath}' because it does not support related documents");
```
Comment (dibarbet) [inline src/LanguageServer/Protocol/Handler/RelatedDocuments/RelatedDocumentsHandler.cs:90]: could consider an early return for readability

File: src/LanguageServer/Protocol/Protocol/Internal/VSInternalRelatedDocumentParams.cs [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+        /// requesting related documents. The server can use this result ID to avoid resending results
+        /// that had previously been sent.
+        /// </summary>
+        [JsonPropertyName("_vs_resultId")]
+        [JsonIgnore(Condition = JsonIgnoreCondition.WhenWritingNull)]
+        public string? ResultId { get; set; }
+
+        [JsonPropertyName("_vs_file_paths")]
+        [JsonIgnore(Condition = JsonIgnoreCondition.WhenWritingNull)]
+        public string[]? FilePaths { get; set; }
```
Comment (dibarbet) [inline src/LanguageServer/Protocol/Protocol/Internal/VSInternalRelatedDocumentParams.cs:42]: seems weird to be using file paths - wondering if we should use a `TextDocumentIdentifier` with a `Uri` - for example what if one if the related docs is a source generated doc?

File: src/LanguageServer/ProtocolUnitTests/RelatedDocuments/RelatedDocumentsTests.cs [hunk 4]
```diff
@@ -0,0 +1,137 @@
...
+        var project = testLspServer.TestWorkspace.CurrentSolution.Projects.Single();
+        var results = await RunGetRelatedDocumentsAsync(
+            testLspServer,
+            project.Documents.First().GetURI(),
+            useProgress: useProgress);
+
+        Assert.Equal(1, results.Length);
+        Assert.Equal(2, results[0].FilePaths!.Length);
+        AssertEx.SetEqual([.. project.Documents.Skip(1).Select(d => d.FilePath)], results[0].FilePaths);
+    }
```
Comment (dibarbet) [inline src/LanguageServer/ProtocolUnitTests/RelatedDocuments/RelatedDocumentsTests.cs:136]: Could use additional test for multiple requests and verify you get the changed / no change responses

File: src/LanguageServer/Protocol/Handler/RelatedDocuments/RelatedDocumentsHandler.cs [hunk 5]
```diff
@@ -0,0 +1,147 @@
...
+    /// Cache where we store the data produced by prior requests so that they can be returned if nothing of significance
+    /// changed. The version key is produced by combining the checksums for project options <see
+    /// cref="ProjectState.GetParseOptionsChecksum"/> and <see cref="DocumentStateChecksums.Text"/>
+    /// </summary>
+    private readonly VersionedPullCache<(Checksum parseOptionsChecksum, Checksum textChecksum)?> _versionedCache = new(nameof(RelatedDocumentsHandler));
+
+    public bool MutatesSolutionState => false;
+    public bool RequiresLSPSolution => true;
+
+    private static async Task<(Checksum parseOptionsChecksum, Checksum textChecksum)> ComputeChecksumsAsync(Document document, CancellationToken cancellationToken)
```
Comment (dibarbet) [inline src/LanguageServer/Protocol/Handler/RelatedDocuments/RelatedDocumentsHandler.cs:47]: Is this endpoint invoked as part of a specific user action, or is the client polling us?

File: src/LanguageServer/ProtocolUnitTests/RelatedDocuments/RelatedDocumentsTests.cs (hunk 4 above)
Comment (CyrusNajmabadi) [reply src/LanguageServer/ProtocolUnitTests/RelatedDocuments/RelatedDocumentsTests.cs:136]: sure.

File: src/LanguageServer/Protocol/Handler/RelatedDocuments/RelatedDocumentsHandler.cs (hunk 2 above)
Comment (CyrusNajmabadi) [reply src/LanguageServer/Protocol/Handler/RelatedDocuments/RelatedDocumentsHandler.cs:90]: sure.

File: src/LanguageServer/Protocol/Protocol/Internal/VSInternalRelatedDocumentParams.cs (hunk 3 above)
Comment (CyrusNajmabadi) [reply src/LanguageServer/Protocol/Protocol/Internal/VSInternalRelatedDocumentParams.cs:42]: @genlu what does copilot want?

File: src/LanguageServer/Protocol/Handler/RelatedDocuments/RelatedDocumentsHandler.cs (hunk 5 above)
Comment (CyrusNajmabadi) [reply src/LanguageServer/Protocol/Handler/RelatedDocuments/RelatedDocumentsHandler.cs:47]: client polling i believe.

File: src/LanguageServer/Protocol/Handler/RelatedDocuments/RelatedDocumentsHandler.cs (hunk 2 above)
Comment (CyrusNajmabadi) [reply src/LanguageServer/Protocol/Handler/RelatedDocuments/RelatedDocumentsHandler.cs:90]: done.

File: src/LanguageServer/Protocol/Protocol/Internal/VSInternalRelatedDocumentParams.cs (hunk 3 above)
Comment (CyrusNajmabadi) [reply src/LanguageServer/Protocol/Protocol/Internal/VSInternalRelatedDocumentParams.cs:42]: So, a source-generated doc won't work at all afaict.  Copilot literally just wants filepaths (they'll point at those on disk).  Seems unfortunate, but it's not a grand design for passing back everything possible.  just likely auxiliary docs they may have already indexed.

**Comment by CyrusNajmabadi** [comment]
@genlu @dibarbet ptal :)

**Review by dibarbet — APPROVED** [review]

**Pull request merged**